src/queue.h

```diff
@@ -27,229 +20,206 @@ typedef struct spsc_queue_header {
+ * @param message_size size (in bytes) of a message. Assume constant message
+ * size for this queue
+ * @return size of usable shm region space
+ **/
+ssize_t create(void* shmaddr,
+               size_t shm_size,
+               size_t message_size) {
+    if (message_size + sizeof(spsc_queue_header) > (u_long) shm_size) {
         perror("Message size type + Header data too large for allocated shm");
         exit(1);
     }
 
-    this->shm = shm;
-
-    //Calculate size of array
-    int num_elements = (shm->size - sizeof(spsc_queue_header)) / sizeof(T);
-    int leftover_bytes = (shm->size - sizeof(spsc_queue_header)) % sizeof(T);
-
+    // Calculate size of array
+    int num_elements = (shm_size - sizeof(spsc_queue_header)) / message_size;
+    int leftover_bytes = (shm_size - sizeof(spsc_queue_header)) % message_size;
 
     spsc_queue_header header;
     header.head = 0;
     header.tail = 0;
+    header.message_size = message_size;
+    header.message_offset = 0;
     header.queue_size = num_elements;
     header.current_count = 0;
     header.total_count = 0;
     header.stop_consumer_polling = false;
     header.stop_producer_polling = false;
-    char* temp_addr = (char*) this->shm->shmaddr ;
+    char* temp_addr = (char*) shmaddr;
     header.message_array = temp_addr + sizeof(spsc_queue_header);
 
-    //Need to do a memcpy? TODO
-    spsc_queue_header* header_ptr = get_queue_header(this->shm->shmaddr);
+    spsc_queue_header* header_ptr = get_queue_header(shmaddr);
     *header_ptr = header;
-    return 0;
+
+    return shm_size - leftover_bytes;
 }
 
 
-template <class T>
-int spsc_queue<T>::attach(shared_memory_region *shm){
-    //do check to verify memory region size and how big to make array considering message size
-    if(sizeof(T) + sizeof(spsc_queue_header) > (u_long) shm->size) {
-        perror("Message size type too large for allocated shm");
-        exit(1);
-    }
-    this->shm = shm;
-    
-    //Need to do a memcpy? TODO
-    // spsc_queue_header* header_ptr = get_queue_header(shm->shmaddr);
-    return 0;
-
+/*
+ * @brief checks if queue is full based on its head and tail positions
+ *
+ * @param header pointer to header of shm queue
+ * @return boolean of whether queue is full
+  */
+bool is_full(spsc_queue_header *header) {
+    return (header->tail + 1) % header->queue_size == header->head;
 }
 
+/*
+ * @brief enqueues buffer to queue
+ *
+ * @param header pointer to queue header
+ * @param buf pointer to buffer we want to enqueue
+ * @param buf_size size of the buf in bytes
+  */
+void enqueue(spsc_queue_header* header, const void* buf, size_t buf_size) {
+    const char* char_buf = (const char*) buf;
+    char* array_start = (char*) header->message_array;
+
+    int message_size = header->message_size;
+    int tail = header->tail;
+
+    // add buf to queue one char byte at a time
+    for (size_t i = 0; i < buf_size; ++i) {
+        array_start[tail*message_size + i] = char_buf[i];
```

Comment:
I am unsure why you need to do this this way. Shouldn't you be able to do a memcopy of the char_buf into the array location? Instead of writing byte by byte?

Reply:
You're totally right! Will make that change in next commit.